src/main/java/problems/impl/TransactionalStack.java

```diff
@@ -0,0 +1,115 @@
+package problems.impl;
+
+import problems.exceptions.TransactionRolledbackException;
+
+import java.util.EmptyStackException;
+import java.util.Stack;
+
+public class TransactionalStack<T> {
+    private final Stack<T> stack = new Stack<>();
+    private final Stack<TransactionEvent<T>> transactionEvents = new Stack<>();
+    private Stack<TransactionEvent<T>> rolledbackTransactionEvents = new Stack<>();
+
+
+    private static class TransactionEvent<E> {
+        private enum Type {
+            BEGIN,
+            PUSH,
+            POP
+        }
+
+        private final Type type;
+        private final E value;
+
+        public TransactionEvent(Type type, E value) {
+            this.type = type;
+            this.value = value;
+        }
+    }
+
+    public void push(T value) {
+        stack.push(value);
+
+        if (!transactionEvents.isEmpty() && rolledbackTransactionEvents.empty()) {
+            transactionEvents.add(new TransactionEvent<>(TransactionEvent.Type.PUSH, value));
+        }
+    }
+
+    public T pop() {
+        T value;
+
+        try {
+            value = stack.pop();
+        } catch (EmptyStackException e) {
+            if (transactionEvents.empty()) {
+                throw e;
+            }
+
+            rollback();
+
+            throw new TransactionRolledbackException("Cannot pop an empty stack. Rolled back changes.");
+        }
+
+        if (!transactionEvents.isEmpty() && rolledbackTransactionEvents.empty()) {
+            transactionEvents.add(new TransactionEvent<>(TransactionEvent.Type.POP, value));
+        }
+
+        return value;
+    }
+
+    public int getSize() {
+        return stack.size();
+    }
+
+    public T peek() {
+        try {
+            return stack.peek();
+        } catch (EmptyStackException e) {
```

Comment:
I don't like this type of code flow (using exception catching to make decisions). It would be better to just check if `stack` is empty before trying to peek, then rollback if you have actions to rollback and last throw the exception yourself.